File: core/entitlements.py
"""
Planes de suscripción (tiers) y mapa de features por tier.

Tiers comerciales: tester (ELIA Tester) y architect (ELIA Architect).
El tier beta desbloquea todo (builds de distribución beta).

Claves legacy BASIC/PRO se normalizan a tester; ENT a architect.
"""
from __future__ import annotations

from typing import Any, Dict, FrozenSet

TIER_TESTER = "tester"
TIER_ARCHITECT = "architect"
TIER_BETA = "beta"

# Slugs legacy (solo lectura / compatibilidad de claves antiguas)
TIER_BASIC = "basic"
TIER_PROFESSIONAL = "professional"
TIER_ENTERPRISE = "enterprise"

TIER_CODES: Dict[str, str] = {
    TIER_TESTER: "TESTER",
    TIER_ARCHITECT: "ARCH",
    TIER_BETA: "BETA",
    # Generación de claves v3 con slugs antiguos
    TIER_BASIC: "BASIC",
    TIER_PROFESSIONAL: "PRO",
    TIER_ENTERPRISE: "ENT",
}

CODE_TO_TIER: Dict[str, str] = {
    "TESTER": TIER_TESTER,
    "ARCH": TIER_ARCHITECT,
    "BETA": TIER_BETA,
    "BASIC": TIER_TESTER,
    "PRO": TIER_TESTER,
    "ENT": TIER_ARCHITECT,
}

_LEGACY_SLUG_TO_CANONICAL: Dict[str, str] = {
    TIER_BASIC: TIER_TESTER,
    TIER_PROFESSIONAL: TIER_TESTER,
    TIER_ENTERPRISE: TIER_ARCHITECT,
}

FEATURE_KEYS: FrozenSet[str] = frozenset(
    {
        "web_recording",
        "api_http_single",
        "doc_to_bdd",
        "api_postman_suites",
        "api_locust",
        "mobile_recording",
        "legacy_recording",
        "publishers_standard",
        "publishers_enterprise",
        "team_memory_crypto",
    }
)

UPGRADE_CONTACT_EMAIL = "elia.qa.software+contacto@gmail.com"


def normalize_tier(tier: str) -> str:
    """Normaliza slug o código legacy al tier canónico (tester | architect | beta)."""
    raw = (tier or "").strip().lower()
    if not raw:
        return ""
    if raw == TIER_BETA:
        return TIER_BETA
    code = raw.upper()
    if code in CODE_TO_TIER:
        return CODE_TO_TIER[code]
    if raw in _LEGACY_SLUG_TO_CANONICAL:
        return _LEGACY_SLUG_TO_CANONICAL[raw]
    if raw in (TIER_TESTER, TIER_ARCHITECT):
        return raw
    return ""


def _base_flags() -> Dict[str, bool]:
    return {key: False for key in FEATURE_KEYS}


def get_tier_flags(tier: str) -> Dict[str, bool]:
    """Devuelve las banderas de características según el tier seleccionado."""
    tier_norm = normalize_tier(tier)
    if not tier_norm:
        return _base_flags()

    if tier_norm == TIER_BETA:
        return {key: True for key in FEATURE_KEYS}

    flags = _base_flags()

    if tier_norm == TIER_TESTER:
        flags["web_recording"] = True
        flags["api_http_single"] = True
        flags["doc_to_bdd"] = True
        flags["api_postman_suites"] = True
        flags["mobile_recording"] = True
        flags["publishers_standard"] = True
        return flags

    if tier_norm == TIER_ARCHITECT:
        for key in FEATURE_KEYS:
            flags[key] = True
        return flags

    return flags


def infer_tier_from_v2_mods(mobile: bool, legacy: bool) -> str:
    """Compatibilidad v2: M/L → tier canónico."""
    if legacy:
        return TIER_ARCHITECT
    return TIER_TESTER


def legacy_modules_from_features(features: Dict[str, bool]) -> Dict[str, bool]:
    """Booleans legacy consumidos por la UI y rutas antiguas."""
    api_any = bool(
        features.get("api_http_single")
        or features.get("api_postman_suites")
        or features.get("api_locust")
    )
    return {
        "mobile_recording": bool(features.get("mobile_recording")),
        "legacy_recording": bool(features.get("legacy_recording")),
        "doc_to_bdd": bool(features.get("doc_to_bdd")),
        "api_testing": api_any,
    }


def tier_display_name(tier: str) -> str:
    tier_norm = normalize_tier(tier) or (tier or "").strip().lower()
    labels = {
        TIER_TESTER: "ELIA Tester",
        TIER_ARCHITECT: "ELIA Architect",
        TIER_BETA: "Beta",
        TIER_BASIC: "ELIA Tester",
        TIER_PROFESSIONAL: "ELIA Tester",
        TIER_ENTERPRISE: "ELIA Architect",
    }
    return labels.get(tier_norm, tier or "—")


def tier_audience(tier: str) -> str:
    tier_norm = normalize_tier(tier)
    if tier_norm == TIER_ARCHITECT:
        return (
            "Bancos, financieras, grandes corporativos y arquitectos de automatización "
            "con infraestructura pesada."
        )
    if tier_norm == TIER_TESTER:
        return (
            "Testers independientes, equipos medianos y células de desarrollo ágil estándar."
        )
    return ""


def tier_required_for_feature(feature: str) -> str:
    """Tier mínimo comercial para upsell UI."""
    tester_features = {
        "doc_to_bdd",
        "api_postman_suites",
        "mobile_recording",
        "publishers_standard",
    }
    if feature in tester_features:
        return TIER_TESTER
    return TIER_ARCHITECT


def upsell_benefits(tier: str) -> list[str]:
    tier_norm = normalize_tier(tier)
    if tier_norm == TIER_TESTER:
        return [
            "Grabación web y cliente HTTP API",
            "Pruebas API en cadena (Postman / suites)",
            "Inteligencia Doc-to-BDD (IA local desde el día uno)",
            "Automatización móvil (Appium)",
            "Publishers Git y Jira Vanilla",
        ]
    if tier_norm == TIER_ARCHITECT:
        return [
            "Todo lo incluido en ELIA Tester",
            "Automatización Legacy (escritorio / pywinauto)",
            "Pruebas de carga Locust, reportes PDF/HTML y SLA",
            "Controladores lógicos (If/Loop), gRPC, JDBC y carga distribuida local",
            "Publishers Xray, Value Edge y Azure DevOps",
            "Team Memory Crypto (seguridad asimétrica local-first)",
        ]
    return []


def entitlements_payload(
    *,
    tier: str,
    features: Dict[str, bool],
    is_beta: bool = False,
    upgrade_email: str = UPGRADE_CONTACT_EMAIL,
) -> Dict[str, Any]:
    tier_norm = normalize_tier(tier) if tier else ""
    legacy = legacy_modules_from_features(features)
    out: Dict[str, Any] = {
        "tier": tier_norm or tier or "",
        "tier_label": tier_display_name(tier_norm or tier),
        "tier_audience": tier_audience(tier_norm or tier),
        "is_beta": is_beta,
        "upgrade_email": upgrade_email,
        "features": dict(features),
        **legacy,
    }
    return out

File: frontend/src/app/entitlements.ts
/** Constantes de tiers (espejo de core/entitlements.py). */

export const TIER_TESTER = "tester";
export const TIER_ARCHITECT = "architect";
export const TIER_BETA = "beta";

/** @deprecated Usar TIER_TESTER */
export const TIER_BASIC = "basic";
/** @deprecated Usar TIER_TESTER */
export const TIER_PROFESSIONAL = "professional";
/** @deprecated Usar TIER_ARCHITECT */
export const TIER_ENTERPRISE = "enterprise";

export const UPGRADE_CONTACT_EMAIL = "elia.qa.software+contacto@gmail.com";

export type FeatureFlags = {
  web_recording?: boolean;
  api_http_single?: boolean;
  api_testing?: boolean;
  doc_to_bdd?: boolean;
  api_postman_suites?: boolean;
  api_locust?: boolean;
  mobile_recording?: boolean;
  legacy_recording?: boolean;
  publishers_standard?: boolean;
  publishers_enterprise?: boolean;
  team_memory_crypto?: boolean;
};

const LEGACY_SLUG_LABELS: Record<string, string> = {
  basic: TIER_TESTER,
  professional: TIER_TESTER,
  enterprise: TIER_ARCHITECT,
};

export function normalizeTier(tier: string): string {
  const raw = (tier || "").trim().toLowerCase();
  if (!raw) return "";
  if (raw === TIER_BETA) return TIER_BETA;
  if (raw in LEGACY_SLUG_LABELS) return LEGACY_SLUG_LABELS[raw];
  if (raw === TIER_TESTER || raw === TIER_ARCHITECT) return raw;
  return "";
}

export function tierDisplayName(tier: string): string {
  const norm = normalizeTier(tier) || tier;
  const labels: Record<string, string> = {
    tester: "ELIA Tester",
    architect: "ELIA Architect",
    beta: "Beta",
    basic: "ELIA Tester",
    professional: "ELIA Tester",
    enterprise: "ELIA Architect",
  };
  return labels[norm] ?? tier;
}

export function tierAudience(tier: string): string {
  const norm = normalizeTier(tier);
  if (norm === TIER_ARCHITECT) {
    return "Bancos, financieras, grandes corporativos y arquitectos de automatización con infraestructura pesada.";
  }
  if (norm === TIER_TESTER) {
    return "Testers independientes, equipos medianos y células de desarrollo ágil estándar.";
  }
  return "";
}

export function upsellBenefits(tier: string): string[] {
  const norm = normalizeTier(tier);
  if (norm === TIER_TESTER) {
    return [
      "Grabación web y cliente HTTP API",
      "Pruebas API en cadena (cliente y suites)",
      "Inteligencia Doc-to-BDD (IA local desde el día uno)",
      "Automatización móvil (Appium)",
      "Publishers Git y Jira Vanilla",
    ];
  }
  if (norm === TIER_ARCHITECT) {
    return [
      "Todo lo incluido en ELIA Tester",
      "Orquestación automática de AVD locales (un clic)",
      "Automatización Legacy (escritorio / pywinauto)",
      "Pruebas de carga Locust, reportes PDF/HTML y SLA",
      "Controladores lógicos (If/Loop), gRPC, JDBC y carga distribuida local",
      "Publishers Xray, Value Edge y Azure DevOps",
      "Team Memory Crypto (seguridad asimétrica local-first)",
    ];
  }
  return [];
}

export function avdOrchestrationAllowed(
  license: { tier?: string | null; is_beta?: boolean } | null | undefined,
): boolean {
  if (!license) return false;
  if (license.is_beta) return true;
  return normalizeTier(license.tier ?? "") === TIER_ARCHITECT;
}

export function hasFeature(modules: FeatureFlags | null | undefined, key: keyof FeatureFlags): boolean {
  if (!modules) return false;
  if (key in modules && typeof modules[key] === "boolean") return Boolean(modules[key]);
  return false;
}

export function featureFromModules(modules: FeatureFlags | null | undefined): FeatureFlags {
  return {
    web_recording: modules?.web_recording ?? true,
    api_http_single: modules?.api_http_single ?? modules?.api_testing ?? false,
    doc_to_bdd: modules?.doc_to_bdd ?? false,
    api_postman_suites: modules?.api_postman_suites ?? false,
    api_locust: modules?.api_locust ?? false,
    mobile_recording: modules?.mobile_recording ?? false,
    legacy_recording: modules?.legacy_recording ?? false,
    publishers_standard: modules?.publishers_standard ?? false,
    publishers_enterprise: modules?.publishers_enterprise ?? false,
    team_memory_crypto: modules?.team_memory_crypto ?? false,
  };
}
